Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,191 @@
+        
+        self.byteBuffer = byteBuffer
+        self.separator = separator
+        self.start = start
+        self.length = length
+        self.currentIndex = self.start
+    }
+
+    public init(byteBuffer: ByteBuffer,
+                separator: UInt8) {
+        
+        self.init(byteBuffer: byteBuffer, separator: separator,
+                  start: byteBuffer.readerIndex, length: byteBuffer.readableBytes)
+    }
+}
+
+extension ByteBuffer {
+    
+    public func sliceByTrimming(whiteSpaces: [UInt8], start: Int, length: Int) -> ByteBuffer? {
+        return withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            assert(start <= self.capacity - length)
+            
+            let firstPtr = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+
+            var startPtr = firstPtr.advanced(by: start)
+            var endPtr   = startPtr.advanced(by: length - 1)
+            
+            // Advance startPtr until not a whiteSpace
+            while startPtr < endPtr && startPtr.pointee != 0 && whiteSpaces.contains(startPtr.pointee) {
+                startPtr = startPtr.advanced(by: 1)
+            }
+            
+            // Retreat endPtr until not a whiteSpace, null is considered a whiteSpace
+            while endPtr >= startPtr && (whiteSpaces.contains(endPtr.pointee) || endPtr.pointee == 0) {
+                endPtr = endPtr.advanced(by: -1)
+            }
+            return self.getSlice(at: startPtr - firstPtr, length: endPtr - startPtr + 1)
+        }
+    }
+    
+    public func sliceByTrimmingWhitespaces() -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: 0, length: self.readableBytes)
+    }
+    
+    public func sliceByTrimmingWhitespaces(from start: Int = 0, length: Int) -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: start, length: length)
+    }
+    
+    public func sliceByTrimmingWhitespaces(from start: Int) -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: start, length: self.readableBytes - start)
+    }
+    
+    /// Compares the buffer to a case insensitive `ContiguousArray<UInt8>`.
+    ///
+    /// This `ContiguousArray` could be get from applying the `asUpperCaseContiguousUTF8UIntArray`
+    ///   property on the string protocol.
+    ///
+    /// **WARNING:** MAKE SURE THAT THE STRING YOU PASS IS UPPERCASE
+    ///
+    /// This function doesn't change the indices of the buffer
+    ///
+    /// - Parameter constant: The string constant in the form of contiguous array _IN UPPER CASE_.
+    /// - Returns: Whether the ByteBuffer contains **EXACTLY** this array or no, but by ignoring case.
+    public func compareReadingToCaseInsensitiveCString(_ constant: ContiguousArray<UInt8>) -> Bool {
```

Comment:
It's probably better to implement this as a generic on top of `Collection<UInt8>` and allow the Swift compiler to handle this, rather than requiring this specificity.